README.md
# Crypto

Helper for web crypto

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.0.4](https://gitlab.com/kohana-js/modules/crypto/compare/v3.0.3...v3.0.4) (2022-09-07)

### 3.0.3 (2022-03-09)

## [3.0.2] - 2021-09-07
### Added
- create CHANGELOG.md

tests/HelperCrypto.test.js
import path from 'node:path';
import HelperCrypto from '../classes/helper/Crypto';

const testKeyHS256 = {
  kty: 'oct', alg: 'HS256', key_ops: ['sign', 'verify'], k: 'ALhJbC1Spav8eSbMJxXe2yR0ZZ-eXneYGaYM51ti1T5pGT9QxYlS6YqQD5Asm4VrVjyTmZLQ00ID1vmn-91ckg', ext: true,
};

import * as url from 'node:url';
const __dirname = url.fileURLToPath(new URL('.', import.meta.url)).replace(/\/$/, '');

describe('test crypto', () => {
  test('make key', async () => {
    const fileName = path.normalize(`${__dirname}/keys/testKey.js`);
    const jwk = await HelperCrypto.makeSignKey(fileName, { name: 'HMAC', hash: 'SHA-256' });
    const {default : key} = await import('./keys/testKey');
    expect(jwk.alg).toBe(key.alg);
    expect(jwk.ext).toBe(key.ext);
    expect(jwk.k).toBe(key.k);
    expect(jwk.kty).toBe(key.kty);

    const fileName2 = path.normalize(`${__dirname}/keys/testKey2.js`);
    const jwk2 = await HelperCrypto.makeSignKey(fileName2);
    const {default: key2} = await import('./keys/testKey2');
    expect(jwk2.alg).toBe(key2.alg);
    expect(jwk2.ext).toBe(key2.ext);
    expect(jwk2.k).toBe(key2.k);
    expect(jwk2.kty).toBe(key2.kty);
  });

  test('sign', async () => {
    const sign = await HelperCrypto.sign(testKeyHS256, 'hello');
    expect(sign).toBe('7NZQDQgE1dV0uu2w0BboV5J7YxTrjhDZQ0+DdXzcIho=');
  });

  test('sign, expire', async () => {
    const expire = 600;
    const timestamp = 1563669060000;
    const sign = await HelperCrypto.sign(testKeyHS256, 'hello', { name: 'HMAC', hash: 'SHA-256' }, expire, timestamp);
    const signs = sign.split('::');
    const signExpire = parseInt(signs[1]);

    expect(sign).toBe('vAMmolmXSfobsltcMc5PmIR4D4QKGAsU84UUYtPL+ro=::1563669660');
    expect(signs[0]).toBe('vAMmolmXSfobsltcMc5PmIR4D4QKGAsU84UUYtPL+ro=');
    expect(signExpire - Math.floor(timestamp / 1000) ).toBe(expire);

    const timestamp2 = 1563669070000;
    const sign2 = await HelperCrypto.sign(testKeyHS256, 'hello', { name: 'HMAC', hash: 'SHA-256' }, expire, timestamp2);
    const sign2s = sign2.split('::');
    const signExpire2 = parseInt(sign2s[1]);
    expect(sign2).toBe('mcUWH87f0GpTnaFdgwDmxTx2CZ+MkHasO6V9xtaD4hA=::1563669670');
    expect(sign2s[0]).toBe('mcUWH87f0GpTnaFdgwDmxTx2CZ+MkHasO6V9xtaD4hA=');
    expect(signExpire2 - Math.floor(timestamp2 / 1000) ).toBe(expire);

    const sign3 = await HelperCrypto.sign(testKeyHS256, 'hello', { name: 'HMAC', hash: 'SHA-256' }, expire);
    const sign3s = sign3.split('::');
    const signExpire3 = parseInt(sign3s[1]);
    expect(signExpire3 - Math.floor(Date.now() / 1000) ).toBe(expire);
  });

  test('verify', async () => {
    const verify = await HelperCrypto.verify(testKeyHS256, '7NZQDQgE1dV0uu2w0BboV5J7YxTrjhDZQ0+DdXzcIho=', 'hello');
    expect(verify).toBe(true);
  });

  test('verify_timestamp', async () => {
    const verify = await HelperCrypto.verify(testKeyHS256, 'mcUWH87f0GpTnaFdgwDmxTx2CZ+MkHasO6V9xtaD4hA=::1563669670', 'hello', { name: 'HMAC', hash: 'SHA-256' }, 1563669060000);
    expect(verify).toBe(true);

    const verify2 = await HelperCrypto.verify(testKeyHS256, 'mcUWH87f0GpTnaFdgwDmxTx2CZ+MkHasO6V9xtaD4hA=::1563669670', 'hello', { name: 'HMAC', hash: 'SHA-256' });
    expect(verify2).toBe(false);
  });

  test('verify false', async () => {
    const verify = await HelperCrypto.verify(testKeyHS256, '7NZQDQgE1dV0uu2w0BboV5J7YxTrjhDZQ0+DdXzcIho=', 'hellos');
    expect(verify).toBe(false);
  });
});


classes/helper/Crypto.mjs
import { Buffer } from 'node:buffer';
import { TextEncoder } from 'node:util';
import { subtle } from 'node:crypto';
import fs from 'node:fs/promises';

const CRYPTO_ALGORITHMS = {
  RSASSA_PKCS1_v1_5: 'RSASSA-PKCS1-v1_5',
  RSA_PSS: 'RSA-PSS',
  ECDSA: 'ECDSA',
  HMAC: 'HMAC',
  RSA_OAEP: 'RSA-OAEP',
  AES_CTR: 'AES-CTR',
  AES_CBC: 'AES-CBC',
  AES_GCM: 'AES-GCM',
  SHA1: 'SHA-1',
  SHA256: 'SHA-256',
  SHA384: 'SHA-384',
  SHA512: 'SHA-512',
  ECDH: 'ECDH',
  HKDF: 'HKDF',
  PBKDF2: 'PBKDF2',
  AES_KW: 'AES-KW',
};

const defaultSignAlgorithm = { name: CRYPTO_ALGORITHMS.HMAC, hash: CRYPTO_ALGORITHMS.SHA256 };

export default class HelperCrypto {
  static CRYPTO_ALGORITHMS = CRYPTO_ALGORITHMS;

  static async makeSignKey(filePath, algorithm = defaultSignAlgorithm) {
    const key = await subtle.generateKey(algorithm, true, ['sign', 'verify']);
    const jwk = await subtle.exportKey('jwk', key);

    await fs.writeFile(filePath, /.json$/.test(filePath)? JSON.stringify(jwk) : `export default ${JSON.stringify(jwk)}`);
    return jwk;
  }

  static async sign(jwk, data, algorithm = defaultSignAlgorithm, expire=0, timestamp= 0) {
    const key = await subtle.importKey('jwk', jwk, algorithm, true, ['sign', 'verify']);
    const expire_ms = expire * 1000;

    const strExpire = (expire ? `::${ Math.floor(((timestamp || Date.now()) + expire_ms) / 1000) }` : '');

    const buffer = new TextEncoder().encode(data + strExpire);
    const sign = await subtle.sign(algorithm, key, buffer);

    return Buffer.from(sign).toString('base64') + strExpire;
  }

  static async verify(jwk, sign, data, algorithm = defaultSignAlgorithm, timestamp=0) {
    const signs = sign.split('::');
    if(signs[1]){
      const expire_ms = parseInt(signs[1]) * 1000;
      if((timestamp || Date.now()) > expire_ms)return false;
    }
    const strExpire = signs[1] ? `::${signs[1]}` : '';

    const key = await subtle.importKey('jwk', jwk, algorithm, true, ['sign', 'verify']);
    const bufferSign = Buffer.from(signs[0], "base64");
    const bufferData = new TextEncoder().encode(data + strExpire);
    return subtle.verify(algorithm, key, bufferSign, bufferData);
  }
}